### PR TITLE
Added Snap Install page to the UI

### LINKED ISSUE #11682: Snaps Installation UI
We should allow dapps to request permission to use snaps when requesting permissions or connecting. If the user doesn't have the snap installed yet, this should initiate the snap install process. We will need to download the snap manifest and present the user with enough information to decide whether or not to install the snap. This should include:

- [x] The snap name
- [x] The snap author
- [x] The place the snap was downloaded from
- [x] The snap description
- [x] The permissions the snap requires
        [ ] This should include detailed description of what each permission is, and why it might matter to a user
- [x] The reason I'm being asked to install the snap (e.g. the dapp that asked permission to use the snap).

This UI should be incorporated into the connect flow, so that the dapp is completely ready to use after the connection process.

As a user, I want to be notified of the following snaps errors:
- When a snap fails to install
  - This will probably be because the snap crashed immediately when MetaMask tried to run it, but it could be for other reasons, so a generic "failed to install" error with a more detailed / verbose error message somehow accessible would be helpful.

https://www.figma.com/file/nBKUQx0btfQiqVf30UA6KD/Flask-MVP?node-id=336%3A9134

### PR DESCRIPTION
**Pull request by ritave**
Closes: #11682 

Screenshot 2021-12-17 at 16 08 34

### CODE REVIEW

File: ui/pages/permissions-connect/permissions-connect.container.js [hunk 1]
```diff
@@ -64,6 +65,11 @@ const mapStateToProps = (state, ownProps) => {
     }
   }
 
+  const isSnap = Boolean(
+    targetSubjectMetadata &&
+      targetSubjectMetadata.subjectType === SUBJECT_TYPES.SNAP,
```
Comment (darkwing) [inline ui/pages/permissions-connect/permissions-connect.container.js:70]: Nit: We can do `targetSubjectMetadata?.subjectType === SUBJECT_TYPES.SNAP` and remove the `targetSubjectMetadata &&`

File: ui/pages/permissions-connect/snap-install/index.scss [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  .snap-requests-permission {
+    @include H7;
+
+    padding: 16px;
+  }
+
+  .permissions-connect-permission-list {
+    padding-left: 24px;
+    padding-right: 24px;
```
Comment (darkwing) [inline ui/pages/permissions-connect/snap-install/index.scss:17]: Do you know if we have padding top or bottom?  If not, we can do `padding: 0 24px;`

File: ui/pages/permissions-connect/snap-install/index.scss [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  .source-code {
+    @include H7;
+
+    display: flex;
+    color: $ui-4;
+
+    .link {
+      color: $primary-blue;
+      margin-left: 4px;
```
Comment (darkwing) [inline ui/pages/permissions-connect/snap-install/index.scss:28]: `margin-inline-start: 4px` would be good for RTL

**Comment by darkwing** [comment]
I'm also seeing `TypeError: can't access property "RestrictedMethod", _snapControllers.PermissionType is undefined`. in the error console

**Comment by ritave** [comment]
> I'm also seeing `TypeError: can't access property "RestrictedMethod", _snapControllers.PermissionType is undefined`. in the error console

This is probably related to Snaps branch in current state, not this PR. I'll check it out in a separate issue

File: ui/components/app/permissions-connect-header/permissions-connect-header.component.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import PropTypes from 'prop-types';
 import React, { Component } from 'react';
-import SiteOrigin from '../../ui/site-origin/site-origin';
+import SiteOrigin from '../../ui/site-origin';
+import SnapsAuthorshipPill from '../flask/snaps-authorship-pill';
```
Comment (Gudahtt) [inline ui/components/app/permissions-connect-header/permissions-connect-header.component.js:4]: This import will fail on non-Flask builds because modules under `flask` directories are excluded. So this will need to be behind a code fence

File: ui/components/app/modals/modal.js [hunk 5]
```diff
@@ -356,6 +359,21 @@ const MODALS = {
     },
   },
 
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  SNAP_INSTALL: {
```
Comment (Gudahtt) [inline ui/components/app/modals/modal.js:363]: For future reference, this entire modal system is deprecated and shouldn't be used for new designs. We'll leave this for now due to our time constraints, but this modal system predates the "React portal" feature, which lets us embed modals directly in the UI next to the code that triggers them, rather than use global state to control them at-a-distance which we had to do previously (because we need them to be rendered at the bottom of the document, so they would be shown above the other content with minimal z-index hacks).

File: ui/components/app/modals/snap-install-warning/snap-install-warning.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  const t = useI18nContext();
+  const dispatch = useDispatch();
+
+  const { onSubmit, snapName } = useSelector(
+    (state) => state.appState.modal.modalState.props,
+  );
+
+  const [isConfirmed, setIsConfirmed] = useState(false);
+
+  const onCheckboxClicked = useCallback(() => setIsConfirmed((i) => !i), []);
```
Comment (Gudahtt) [inline ui/components/app/modals/snap-install-warning/snap-install-warning.js:18]: Nit: We should use a more descriptive variable name here

```suggestion
  const onCheckboxClicked = useCallback(() => setIsConfirmed((confirmedState) => !confirmedState), []);
```

File: ui/components/app/modals/snap-install-warning/snap-install-warning.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      cancelText={t('cancel')}
+      submitDisabled={!isConfirmed}
+    >
+      <div className="snap-install-warning">
+        <div className="header">{t('areYouSure')}</div>
+        <div className="subheader">{t('snapInstallWarningCheck')}</div>
+        <div className="checkbox-label">
+          <CheckBox
+            checked={isConfirmed}
+            disabled={false}
```
Comment (Gudahtt) [inline ui/components/app/modals/snap-install-warning/snap-install-warning.js:34]: Nit: This is the default for this prop, so we can omit it

```suggestion
```

File: ui/components/app/modals/snap-install-warning/snap-install-warning.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  return (
+    <Modal
+      onSubmit={onSubmit}
+      onCancel={onCancel}
+      submitText={t('approveAndInstall')}
+      cancelText={t('cancel')}
+      submitDisabled={!isConfirmed}
+    >
+      <div className="snap-install-warning">
+        <div className="header">{t('areYouSure')}</div>
```
Comment (darkwing) [inline ui/components/app/modals/snap-install-warning/snap-install-warning.js:29]: We should use our `Typography` component here:

```
<Typography variant={TYPOGRAPHY.H7} fontWeight={FONT_WEIGHT.BOLD}
```

File: ui/components/app/modals/snap-install-warning/snap-install-warning.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        <div className="subheader">{t('snapInstallWarningCheck')}</div>
```
Comment (darkwing) [inline ui/components/app/modals/snap-install-warning/snap-install-warning.js:30]: Same here:  

```
<Typography variant={TYPOGRAPHY.H6} paddingLeft={16} paddingRight={16}
```

**Review by darkwing — CHANGES_REQUESTED** [review]

File: ui/pages/permissions-connect/permissions-connect.container.js [hunk 10]
```diff
@@ -54,6 +57,10 @@ const mapStateToProps = (state, ownProps) => {
     subjectType: SUBJECT_TYPES.UNKNOWN,
   };
 
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  const isSnap = targetSubjectMetadata?.subjectType === SUBJECT_TYPES.SNAP;
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.container.js:61]: Nit: `targetSubjectMetadata` should always be an object (see the `??` operator on the above line), so this `?` isn't needed

```suggestion
  const isSnap = targetSubjectMetadata.subjectType === SUBJECT_TYPES.SNAP;
```

File: ui/pages/permissions-connect/permissions-connect.container.js [hunk 11]
```diff
@@ -68,18 +75,29 @@ const mapStateToProps = (state, ownProps) => {
...
+  ///: END:ONLY_INCLUDE_IN
 
   let page = '';
   if (pathname === connectPath) {
     page = '1';
   } else if (pathname === confirmPermissionPath) {
     page = '2';
+    ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  } else if (pathname === snapInstallPath) {
+    page = '3';
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.container.js:89]: This doesn't seem right. The total number of pages is still hard-coded as 2 (in `permissions-connect.component.js`) for starters, so the header would show "page 3 / 2" here.

I would expect that the total number of pages would be 1 as a baseline, plus 1 if `eth_accounts` was requested, and plus 1 if a snap was requested that hasn't been installed yet. Then calculating the _current_ page depends on that total. This should only be 3 if `eth_accounts` is also being requested, otherwise this would be 2.

**Comment by Gudahtt** [comment]
How have you been testing this by the way? The "snaps template" website didn't seem to work with this branch.

**Comment by ritave** [comment]
> How have you been testing this by the way? The "snaps template" website didn't seem to work with this branch.

Just tried it again - works for me.

File: ui/components/app/modals/modal.js (hunk 5 above)
Comment (ritave) [reply ui/components/app/modals/modal.js:363]: Do you have any code snippets on how to build modals in the future?

File: ui/store/actions.js [hunk 12]
```diff
@@ -655,6 +655,20 @@ export function updateCustomNonce(value) {
   };
 }
 
+///: BEGIN:ONLY_INCLUDE_IN(flask)
+export function showSnapInstallWarning(snapName, onSubmit) {
+  return (dispatch) => {
+    dispatch(
+      showModal({
+        name: 'SNAP_INSTALL',
+        onSubmit,
```
[the remark this answers is not in the document]
Comment (ritave) [reply ui/store/actions.js:664]: This is done by other modals, eg. `ConfirmDeleteNetwork`.

File: ui/pages/permissions-connect/snap-install/snap-install.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                })
+              }
+            >
+              {t('openSourceCode')}
+            </div>
+          </div>
+        ) : null}
+        <PermissionsConnectFooter />
+        <PageContainerFooter
+          cancelButtonTyype="default"
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/snap-install/snap-install.js:89]: Typo:
```suggestion
          cancelButtonType="default"
```

File: ui/pages/permissions-connect/snap-install/snap-install.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+  return (
+    <div className="page-container snap-install">
+      <div className="headers">
+        <PermissionConnectHeader
+          icon={targetSubjectMetadata.iconUrl}
+          iconName={targetSubjectMetadata.name}
+          headerTitle={t('snapInstall')}
+          headerText={null} // TODO(ritave): Add header text when snaps support description
+          siteOrigin={targetSubjectMetadata.name}
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/snap-install/snap-install.js:63]: The closest thing we have to an origin is the npm ID I think? The name certainly seems like the wrong thing to show here.

File: ui/pages/permissions-connect/snap-install/snap-install.js [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+          </div>
+        ) : null}
+        <PermissionsConnectFooter />
+        <PageContainerFooter
+          cancelButtonTyype="default"
+          onCancel={onCancel}
+          cancelText={t('cancel')}
+          onSubmit={shouldShowWarning ? showWarning : onSubmit}
+          submitText={t('approveAndInstall')}
+          buttonsSizeLarge={false}
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/snap-install/snap-install.js:94]: This is a typo as well (the prop is `buttonSizeLarge`, but also `false` is the default so we don't need to include this at all

```suggestion
```

File: ui/pages/permissions-connect/snap-install/snap-install.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+          headerText={null} // TODO(ritave): Add header text when snaps support description
+          siteOrigin={targetSubjectMetadata.name}
+          npmPackageName={npmId}
+        />
+        <div className="snap-requests-permission">
+          {t('snapRequestsPermission')}
+        </div>
+        <PermissionsConnectPermissionList permissions={request.permissions} />
+      </div>
+      <div className="footers">
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/snap-install/snap-install.js:71]: Nit: The format we use for CSS classes is BEM - Block, Element, Modifier. See here for details: http://getbem.com/introduction/

The basic idea is that any component that makes sense on its own, as an independent component, is a "block", and each sub-piece that is a part of a greater whole is an element. And modifiers are for temporary states like "disabled", often not used.

So for the class names in this component to be BEM-compliant, this would be something like `snap-install__footer`.

Not a blocker, these class names work as-is functionally. Just wanted to make a note for future reference at least though, so your class names can be in-line with those used elsewhere in the wallet.

File: ui/components/app/modals/modal.js (hunk 5 above)
Comment (Gudahtt) [reply ui/components/app/modals/modal.js:363]: Yep, we have a dedicated component for this: `Popover`. There are a few examples of how this is used throughout the codebase.

File: ui/pages/permissions-connect/snap-install/snap-install.js (hunk 14 above)
Comment (ritave) [reply ui/pages/permissions-connect/snap-install/snap-install.js:63]: This `siteOrigin` is probably not the best name for this property. It's displayed in a Chip as a name of the Snap

File: ui/pages/permissions-connect/permissions-connect.container.js [hunk 17]
```diff
@@ -68,18 +75,30 @@ const mapStateToProps = (state, ownProps) => {
 
   const connectPath = `${CONNECT_ROUTE}/${permissionsRequestId}`;
   const confirmPermissionPath = `${CONNECT_ROUTE}/${permissionsRequestId}${CONNECT_CONFIRM_PERMISSIONS_ROUTE}`;
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  const snapInstallPath = `${CONNECT_ROUTE}/${permissionsRequestId}${CONNECT_SNAP_INSTALL_ROUTE}`;
+  ///: END:ONLY_INCLUDE_IN
 
+  const totalPages = (1 + isRequestingAccounts + isSnap).toString();
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.container.js:82]: This crashes the UI in a normal build because `isSnap` is not defined. So we need a fence here, for the Flask and non-Flask version of this calculation. Or perhaps it would be simpler to remove the fence around the `isSnap` variable.

Comment (Gudahtt) [reply ui/pages/permissions-connect/permissions-connect.container.js:82]: Also, adding a number to a boolean is unintuitive. Maybe it would be simpler for readability to do this the more verbose way.  e.g. something like:

```
const totalPages = 1 + (isRequestingAccounts ? 1 : 0) + (isSnap ? 1 : 0);
```

File: ui/pages/permissions-connect/snap-install/snap-install.js (hunk 14 above)
Comment (Gudahtt) [reply ui/pages/permissions-connect/snap-install/snap-install.js:63]: On second thought, I don't think the snap should be shown in the "SiteOrigin" component at all. That design would be inconsistent, and might make it easier to mislead users with misleading snap names. I created a thread about this on Slack.

**Comment by Gudahtt** [comment]
I was able to test this with the "snaps template" website after, I think I must have been using the wrong build type on my last attempt. That template doesn't request key access though, so the warning is not triggered. Do you have any suggestions for testing that warning? Or should I go ahead and clone that template locally and modify it.

**Comment by darkwing** [comment]
From a design/user perspective, `snap_confirm` doesn't really mean anything to me:
Install

I believe we should maybe bold `snap_confirm` and then underneath it, provide a description of what exactly that means, what the extent of the permission does, etc.  Otherwise users find themselves in a spot where it's all or nothing, aka "yolo" or "oh no".

File: ui/pages/permissions-connect/permissions-connect.container.js [hunk 18]
```diff
@@ -68,18 +75,35 @@ const mapStateToProps = (state, ownProps) => {
...
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  totalPages += isSnap;
+  ///: END:ONLY_INCLUDE_IN
+  totalPages = totalPages.toString();
 
   let page = '';
   if (pathname === connectPath) {
     page = '1';
   } else if (pathname === confirmPermissionPath) {
     page = '2';
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.container.js:92]: If no `eth_accounts` permission is requested, this will be the first page:

```suggestion
    page = isRequestingAccounts ? '2' : '1';
```

File: ui/pages/permissions-connect/permissions-connect.component.js [hunk 19]
```diff
@@ -188,7 +208,7 @@ export default class PermissionConnect extends Component {
         ) : null}
         {isRequestingAccounts ? (
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.component.js:209]: Why is this hidden if no accounts are requested? :thinking: This results in the page number being hidden for the snaps install flow as well.

Comment (Gudahtt) [reply ui/pages/permissions-connect/permissions-connect.component.js:209]: The "back" button should be shown on the snap install screen as well, which would require the `goBack` function being updated to handle either page 2 or 3.

File: ui/pages/permissions-connect/snap-install/snap-install.js [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+        {targetSubjectMetadata.sourceCode ? (
+          <div className="source-code">
+            <div className="text">{t('areYouDeveloper')}</div>
+            <div
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/snap-install/snap-install.js:75]: Nit: we should be using anchors or buttons for clickable elements, rather than divs. Using built-in elements that are interactive from the start makes it easier to ensure the wallet is accessible via keyboard, screen reader, and other accessibility tools.

**Comment by rekmarks** [comment]
@Gudahtt @darkwing @ritave re: the `snap_confirm` text, on the base branch (`snaps`) the permissions descriptions are added correctly so whatever is going on in @darkwing's screenshot shouldn't be happening. Ref:

File: ui/pages/permissions-connect/permissions-connect.component.js (hunk 19 above)
Comment (ritave) [reply ui/pages/permissions-connect/permissions-connect.component.js:209]: The back button is a harder thing to implement - it would require removing already approved permissions, due to time constraints I suggest we leave it without a back button.

**Comment by ritave** [comment]
> @Gudahtt @darkwing @ritave re: the `snap_confirm` text, on the base branch (`snaps`) the permissions descriptions are added correctly so whatever is going on in @darkwing's screenshot shouldn't be happening. Ref:
> 
> 

This shouldn't, and doesn't happen when I test it

**Review by Gudahtt — COMMENTED** [review]
When I press 'cancel' on the snap install page, the UI crashes with this error:

```
Uncaught TypeError: Cannot convert undefined or null to object
  at Function.keys (<anonymous>)
  at ui-2.js:15311
  at updateMemo (ui-0.js:35031)
  at Object.useMemo (ui-0.js:35511)
  at useMemo (common-5.js:30162)
  at SnapInstall (ui-2.js:15311)
  at renderWithHooks (ui-0.js:34437)
  at updateFunctionComponent (ui-0.js:36524)
  at beginWork$1 (ui-0.js:38353)
  at HTMLUnknownElement.callCallback (ui-0.js:18513)
backend.js:12979 The above error occurred in the <SnapInstall> component:
    in SnapInstall (created by Context.Consumer)
    in Route (created by PermissionConnect)
    in Switch (created by PermissionConnect)
    in div (created by PermissionConnect)
    in PermissionConnect (created by ConnectFunction)
    in ConnectFunction (created by Context.Consumer)
    in Route (created by Authenticated)
    in Authenticated (created by ConnectFunction)
    in ConnectFunction (created by Routes)
    in Switch (created by Routes)
    in div (created by Routes)
    in div (created by Routes)
    in Routes (created by ConnectFunction)
    in ConnectFunction (created by Context.Consumer)
    in withRouter(Connect(Routes)) (created by Index)
    in LegacyI18nProvider (created by Index)
    in I18nProvider (created by Index)
    in LegacyMetaMetricsProvider (created by Index)
    in MetaMetricsProvider (created by Index)
    in LegacyMetaMetricsProvider (created by Index)
    in MetaMetricsProvider (created by Index)
    in Router (created by HashRouter)
    in HashRouter (created by Index)
    in Provider (created by Index)
    in Index

React will try to recreate this component tree from scratch using the error boundary you provided, Index.
```

File: ui/pages/permissions-connect/permissions-connect.component.js (hunk 19 above)
Comment (Gudahtt) [reply ui/pages/permissions-connect/permissions-connect.component.js:209]: I hadn't realized that the snap installation was handled by a separate permission request. I agree that postponing the pagination makes sense for now, including the back functionality.

**Comment by ritave** [comment]
> When I press 'cancel' on the snap install page, the UI crashes with this error:
> 
> ```
> Uncaught TypeError: Cannot convert undefined or null to object
>   at Function.keys (<anonymous>)
>   at ui-2.js:15311
>   at updateMemo (ui-0.js:35031)
>   at Object.useMemo (ui-0.js:35511)
>   at useMemo (common-5.js:30162)
>   at SnapInstall (ui-2.js:15311)
>   at renderWithHooks (ui-0.js:34437)
>   at updateFunctionComponent (ui-0.js:36524)
>   at beginWork$1 (ui-0.js:38353)
>   at HTMLUnknownElement.callCallback (ui-0.js:18513)
> backend.js:12979 The above error occurred in the <SnapInstall> component:
>     in SnapInstall (created by Context.Consumer)
>     in Route (created by PermissionConnect)
>     in Switch (created by PermissionConnect)
>     in div (created by PermissionConnect)
>     in PermissionConnect (created by ConnectFunction)
>     in ConnectFunction (created by Context.Consumer)
>     in Route (created by Authenticated)
>     in Authenticated (created by ConnectFunction)
>     in ConnectFunction (created by Routes)
>     in Switch (created by Routes)
>     in div (created by Routes)
>     in div (created by Routes)
>     in Routes (created by ConnectFunction)
>     in ConnectFunction (created by Context.Consumer)
>     in withRouter(Connect(Routes)) (created by Index)
>     in LegacyI18nProvider (created by Index)
>     in I18nProvider (created by Index)
>     in LegacyMetaMetricsProvider (created by Index)
>     in MetaMetricsProvider (created by Index)
>     in LegacyMetaMetricsProvider (created by Index)
>     in MetaMetricsProvider (created by Index)
>     in Router (created by HashRouter)
>     in HashRouter (created by Index)
>     in Provider (created by Index)
>     in Index
> 
> React will try to recreate this component tree from scratch using the error boundary you provided, Index.
> ```

Fixed

File: ui/helpers/utils/transactions.util.js [hunk 21]
```diff
@@ -212,7 +212,7 @@ export function getTransactionTypeTitle(t, type, nativeCurrency = 'ETH') {
       return t('transferFrom');
     }
     case TRANSACTION_TYPES.TOKEN_METHOD_APPROVE: {
-      return t('approve');
+      return t('approveLimit');
```
Comment (Gudahtt) [inline ui/helpers/utils/transactions.util.js:215]: This change seems unrelated to the snap install page. A few of the locale changes related to approving seem unrelated as well. Merge conflict resolution mistake perhaps?

Comment (ritave) [reply ui/helpers/utils/transactions.util.js:215]: It's a fix for snap-confirm.js having wrong button text

File: ui/pages/permissions-connect/permissions-connect.component.js [hunk 22]
```diff
@@ -267,6 +290,29 @@ export default class PermissionConnect extends Component {
...
             />
+            {
+              ///: BEGIN:ONLY_INCLUDE_IN(flask)
+            }
+            <Route
+              path={snapInstallPath}
+              exact
+              render={() => (
+                <SnapInstall
+                  request={permissionsRequest || {}}
```
Comment (Gudahtt) [inline ui/pages/permissions-connect/permissions-connect.component.js:301]: `permissionsRequest` should not be empty :thinking: At least not unless `redirecting` is `true`, in which case this would not be rendered. See the `componentDidUpdate` lifecycle function

File: ui/helpers/utils/transactions.util.js (hunk 21 above)
Comment (Gudahtt) [reply ui/helpers/utils/transactions.util.js]: Ahh I see, OK thanks!

File: ui/pages/permissions-connect/permissions-connect.component.js (hunk 22 above)
Comment (ritave) [reply ui/pages/permissions-connect/permissions-connect.component.js:301]: Check PermissionPageContainer route, the same safeguard is there in place.

**Review by Gudahtt — APPROVED** [review]
LGTM! A few things to improve still but this seems to work

**Review by darkwing — APPROVED** [review]
